Reply type decorator

aqk opened the pull request:
Note that part of this depends on calling the named protocol functions, and part of the checks are in `send_message` / `send_to_all`, `send_to_all_except`

When / before Almog lands https://github.com/Chia-Network/chia-blockchain/pull/8093/ I can submit a PR to his branch to join the two approaches (if possible).

aqk committed:
Check types of messages returned from other peers

aqk committed:
Check message type data structures at startup, check that no peeer messages that expect replies are broadcast, and dynamically check return types of messages that do expect replies.

aqk committed:
Type hint protocol message type check functions

aqk committed:
lint

aqk committed:
typo

almogdepaz committed:
add replay_type decorator

almogdepaz committed:
add api reply decorations

almogdepaz committed:
typo

aqk committed:
Don't check non full-node broadcast messages

arvidn commented on chia/util/api_decorators.py:
```diff
@@ -59,3 +59,14 @@ def inner():
+def reply_type(type):
+    def wrap(func):
+        def inner():
+            setattr(func, "reply_type", type)
```
is this attribute used anywhere?

aqk replied:
Almog uses this in #8093 - in that patch, he is building the expected message return types into attributes on the sending functions.

arvidn replied:
I see

aqk committed:
Distinguish internal vs peer protocol error
Added Changelog
Moved static check to import time
Moved protocol timeouts to chia/protocols/protocol_timing.py
Comment typos
Rename create_request -> send_request
Comment that INVALID_PROTOCOL_MESSAGE is bannable, not temporary

aqk commented on chia/protocols/protocol_state_machine.py:
```diff
@@ -58,3 +58,8 @@ def sent_message_response_ok(sent: ProtocolMessageTypes, received: Optional[Prot
+
+def main():
+    # Check this static invariant at import time
+    static_check_sent_message_response()
```
The need for this will go away when we remove `VALID_REPLY_MESSAGE_MAP`, and use only the `reply_type` decerator.

arvidn replied:
does anyone call `main()`? The comment seem to suggest that it's called at import time, but I don't see it.
Did you mean to put this call in the global scope?

aqk commented on chia/protocols/protocol_timing.py:
```diff
@@ -0,0 +1,4 @@
+# These settings should not be end-user configurable
+INVALID_PROTOCOL_BAN_SECONDS = 10
```
I stuck these in here because I didn't want users fiddling with them as config settings.

aqk commented on chia/server/server.py:
```diff
@@ -615,7 +616,14 @@ async def send_to_all(self, messages: List[Message], node_type: NodeType):
+                        # Internal peer protocol logic error
+                        # If we raise ProtocolError here exiting the loop will affect other peers
+                        self.log.error(f"send_to_all called with message expecting a response: {message.type}")
+                        await connection.close(
```
These two calls to `connection.close` with `WSCloseCode.INTERNAL_ERROR` are because we do not expect to call `send_message` with peer messages that expect a reply.

See `chia/server/ws_connection.py` send_message

arvidn replied:
if we accidentally do this, and it's triggered by a test, we won't necessarily fail the test, because tests are allowed to print errors to the log

aqk replied:
@arvidn @almogdepaz What's a good solution here?

Should we save the fact that we had an error, and throw at the end of the loop?
Assert during tests?

aqk replied:
Ah, you know what, I can raise here, because the check is on the message type - it would affect all clients.

arvidn replied:
I think `raise` would be reasonable

aqk commented on chia/server/ws_connection.py:
```diff
@@ -272,14 +279,20 @@ async def invoke(*args, **kwargs):
             msg: Message = Message(uint8(getattr(ProtocolMessageTypes, attr_name).value), None, args[0])
             request_start_t = time.time()
-            result = await self.create_request(msg, timeout)
+            result = await self.send_request(msg, timeout)
```
unrelated name change - seems more clear

aqk commented on chia/util/errors.py:
```diff
@@ -7,7 +7,7 @@ class Err(Enum):
     BAD_HEADER_SIGNATURE = -2
     MISSING_FROM_STORAGE = -3
-    INVALID_PROTOCOL_MESSAGE = -4
+    INVALID_PROTOCOL_MESSAGE = -4  # We WILL ban for a protocol violation.
```
Note the comment on line 7:

`# temporary errors. Don't blacklist`

aqk commented:
6580b3b bans peers for protocol violations, and also distinguishes a protocol error from our client (logic error) vs. a protocol reply error (peer bug or DOS)

arvidn commented on chia/protocols/protocol_state_machine.py:
```diff
@@ -0,0 +1,65 @@
+        raise AssertionError("Overlapping NO_REPLY_EXPECTED and VAILD_REPLY_MESSAGE_MAP values: {}")
+
+
+def sent_message_response_ok(sent: ProtocolMessageTypes, received: Optional[ProtocolMessageTypes]) -> bool:
```
this name is a bit confusing to me. I would expect it to be called `message_response_ok()`, `validate_response()` or something along those lines.

It's checking whether the response is OK, right? the `sent`-prefix, indicating that it's a response to a message we *sent* seems redundant. Clearly we wouldn't receive a response to a message we *didn't* send.

aqk replied:
Right, renaming this

arvidn reviewed (commented):
is there a way we can test this?

aqk committed:
Call static check at module import time

aqk committed:
Rename message_response_ok

aqk committed:
Improve protocol checking for outgoing message validation

aqk committed:
Type-o

aqk committed:
reset submodule

wjblanke committed:
Merge branch 'main' into replay_type_decorator

wjblanke reviewed (approved):
aok
